docs/proposals/rescheduling.md

```diff
@@ -0,0 +1,496 @@
+  (controllers, schedulers, an entirely new component e.g. "rescheduler", etc.)
+* rescheduler design issues: how much does a rescheduler need to know about pods'
+  schedulers' policies, how does the rescheduler specify its rescheduling
+  requests/decisions (e.g. just as an eviction, an eviction with a hint about where to
+  reschedule, or as an eviction paired with a specific binding), how does the system
+  implement these requests, does the rescheduler take into account the second-order
+  effects of decisions (e.g. whether an evicted pod will reschedule, will cause
+  a preemption when it reschedules, etc.), does the rescheduler execute multi-step plans
+  (e.g. evict two pods at the same time with the intent of moving one into the space
+  vacated by the other, or even more complex plans)
+
+Additional musings on the rescheduling design space can be found [here](rescheduler.md).
+
+## Design proposal
+
+The key mechanisms and components of the proposed design are priority, preemption,
+disruption budgets, the `/evict` subresource, and the rescheduler.
+
+### Priority
+
+#### Motivation
+
+[Quality of Service levels](resource-qos.md) allow Kubernetes to safely overcommit nodes.
+Just as it is useful to overcommit nodes to increase node-level utilization, it is useful
+to overcommit clusters to increase cluster-level utilization.  Scheduling priority (which
+we abbreviate as *priority*, in combination with disruption budgets (described in the
+next section), allows Kubernetes to safely overcommit clusters much as QoS levels allow
+it to safely overcommit nodes.
+
+Today, cluster sharing among users, workload types, etc. is regulated via the
+[quota](../admin/resourcequota/README.md) mechanism. When allocating quota, a cluster
+administrator has two choices: (1) the sum of the quotas is less than or equal to the
+capacity of the cluster, or (2) the sum of the quotas is greater than the capacity of the
+cluster (that is, the cluster is overcommitted).  (1) is likely to lead to cluster
+under-utilization, while (2) is unsafe in the sense that someone's pods may go pending
+indefinitely even though they are still within their quota. Priority makes cluster
+overcommitment (i.e. case (2)) safe by allowing users and/or administrators to identify
+which pods should be allowed to run, and which should go pending, when demand for cluster
+resources exceeds supply to due to cluster overcommitment.
+
+Priority is also useful in some special-cause scenarios, such as ensuring that system
+DaemonSets can always schedule and reschedule onto every node where they want to run
+(assuming they are given the highest priority), e.g. see #21767.
+
+#### Specifying priorities
+
+We propose to add a required `Priority` field to `PodSpec`. Its value type is string, and
+the cluster administrator defines a total ordering on these strings (for example
+`Critical`, `Normal`, `Preemptible`). We choose string instead of integer so that it is
```

Comment:
Thought about this again on my ride in this morning.  We know that exposing a numerical ordering is messy, but with this UI, users will have to think in terms of a named priority (hopefully they remember the semantics/ordering!) and the QoS behavior they get based on the quantitative rules that map to conceptual words documented [here](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/resource-qos.md).

It's true that we don't want to expose numbers to users, but what we'd like to expose (IMO) is fully-packaged Service Levels with names.  They *aren't* ordered themselves because they're multi-axis.  "Uncheckpointable Batch", for example, would map to specific priority and QoS settings.

If we plan on going this direction, we should consider not forcing users to name the priorities and sticking with numbers, because ultimately the numbers should be hidden behind the Service Level facade and naming the objects in the total ordering is actually going to be [hard and error-prone](http://martinfowler.com/bliki/TwoHardThings.html).